### PR TITLE
UCT/IB/RC: Scatter to CQE upstream support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8300/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8301/ for details (Mellanox internal link).

**Comment by brminich** [comment]
arm env issues

bot:mlx:retest

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 1]
```diff
@@ -365,7 +395,7 @@ static uct_rc_iface_ops_t uct_rc_mlx5_iface_ops = {
     .event_cq                 = uct_rc_mlx5_iface_event_cq,
     .handle_failure           = uct_rc_mlx5_iface_handle_failure,
     .set_ep_failed            = uct_rc_mlx5_ep_set_failed,
-    .create_qp                = uct_ib_iface_create_qp
+    .create_qp                = uct_rc_mlx5_iface_create_qp
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: Isn't it applicable for DC as well?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8318/ for details (Mellanox internal link).

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 1 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: DC is always signaled, so it's always scatters

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8349/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 2]
```diff
@@ -179,6 +179,36 @@ static void uct_rc_mlx5_iface_progress_enable(uct_iface_h tl_iface, unsigned fla
                                       iface->super.progress, flags);
 }
 
+static ucs_status_t uct_rc_mlx5_iface_create_qp(uct_ib_iface_t *iface,
+                                                uct_ib_qp_attr_t *attr,
+                                                struct ibv_qp **qp_p)
+{
+#if HAVE_DECL_MLX5DV_QP_CREATE_ALLOW_SCATTER_TO_CQE
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: let's always use mlx5dv_create_qp for rc_mlx5, if mlx5dv_create_qp  exists.
then, if HAVE_DECL_MLX5DV_QP_CREATE_ALLOW_SCATTER_TO_CQE, pass this flag.
(i'm thinking about other reasons we might want to use mlx5dv_create_qp )

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 3]
```diff
@@ -179,6 +179,36 @@ static void uct_rc_mlx5_iface_progress_enable(uct_iface_h tl_iface, unsigned fla
...
+    uct_ib_device_t *dev               = uct_ib_iface_device(iface);
+    struct mlx5dv_qp_init_attr dv_attr = {};
+    struct ibv_qp *qp;
+
+    uct_ib_iface_fill_attr(iface, attr);
+
+    if (attr->sq_sig_all) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: why do we need that? mlx5 transport code should make sure all potential sc2cqe posts are signaled

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 4]
```diff
@@ -179,6 +179,36 @@ static void uct_rc_mlx5_iface_progress_enable(uct_iface_h tl_iface, unsigned fla
...
+        dv_attr.comp_mask    = MLX5DV_QP_INIT_ATTR_MASK_QP_CREATE_FLAGS;
+        dv_attr.create_flags = MLX5DV_QP_CREATE_ALLOW_SCATTER_TO_CQE;
+    }
+    qp = mlx5dv_create_qp(dev->ibv_context, &attr->ibv, &dv_attr);
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: space line before

**Comment by yosefe** [comment]
replaced by #3065

**Comment by yosefe** [comment]
@brminich @Artemy-Mellanox yes! sorry for confusion

**Pull request closed without merging**